cpp/include/raft/sparse/solver/detail/lanczos.cuh

```diff
@@ -1396,4 +1438,658 @@ int computeLargestEigenvectors(
+    if (row > 0) { M[(row - 1) * n + row] = beta[row - 1]; }
+  }
+}
+
+template <typename T>
+RAFT_KERNEL kernel_triangular_beta_k(T* t, const T* beta_k, int k, int n)
+{
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  if (tid < k) {
+    // Update the k-th column: t[i, k] -> t[k * n + i] in column-major
+    t[tid * n + k] = beta_k[tid];
+
+    // Update the k-th row: t[k, j] -> t[j * n + k] in column-major
+    t[k * n + tid] = beta_k[tid];
+  }
+}
+
+template <typename T>
+RAFT_KERNEL kernel_normalize(const T* u, const T* beta, int j, int n, T* v, T* V, int size)
+{
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (i < size) {
+    if (beta[j] == 0) {
+      v[i] = u[i] / 1;
+    } else {
+      v[i] = u[i] / beta[j];
+    }
+    V[i + (j + 1) * n] = v[i];
+  }
+}
+
+template <typename T>
+RAFT_KERNEL kernel_clamp_down(T* value, T threshold)
+{
+  *value = (fabs(*value) < threshold) ? 0 : *value;
+}
+
+template <typename T>
+RAFT_KERNEL kernel_clamp_down_vector(T* vec, T threshold, int size)
+{
+  int idx = threadIdx.x + blockIdx.x * blockDim.x;
+  if (idx < size) { vec[idx] = (fabs(vec[idx]) < threshold) ? 0 : vec[idx]; }
+}
+
+template <typename index_type_t, typename value_type_t>
+void cupy_solve_ritz(
+  raft::resources const& handle,
+  raft::device_matrix_view<value_type_t, uint32_t, raft::row_major> alpha,
+  raft::device_matrix_view<value_type_t, uint32_t, raft::row_major> beta,
+  std::optional<raft::device_vector_view<value_type_t, uint32_t, raft::row_major>> beta_k,
+  index_type_t k,
+  int which,
+  int ncv,
+  raft::device_matrix_view<value_type_t, uint32_t, raft::col_major> eigenvectors,
+  raft::device_vector_view<value_type_t> eigenvalues)
+{
+  auto stream = resource::get_cuda_stream(handle);
+
+  value_type_t zero = 0;
+  auto triangular_matrix =
+    raft::make_device_matrix<value_type_t, uint32_t, raft::col_major>(handle, ncv, ncv);
+  raft::matrix::fill(handle, triangular_matrix.view(), zero);
+
+  raft::matrix::initializeDiagonalMatrix(
+    alpha.data_handle(), triangular_matrix.data_handle(), ncv, ncv, stream);
+
+  int blockSize = 256;
+  int numBlocks = (ncv + blockSize - 1) / blockSize;
+  kernel_triangular_populate<value_type_t>
+    <<<blockSize, numBlocks>>>(triangular_matrix.data_handle(), beta.data_handle(), ncv);
+
+  if (beta_k) {
+    int threadsPerBlock = 256;
+    int blocksPerGrid   = (k + threadsPerBlock - 1) / threadsPerBlock;
+    kernel_triangular_beta_k<value_type_t><<<blocksPerGrid, threadsPerBlock>>>(
+      triangular_matrix.data_handle(), beta_k.value().data_handle(), (int)k, ncv);
+  }
+
+  auto triangular_matrix_view =
+    raft::make_device_matrix_view<const value_type_t, uint32_t, raft::col_major>(
+      triangular_matrix.data_handle(), ncv, ncv);
+
+  raft::linalg::eig_dc(handle, triangular_matrix_view, eigenvectors, eigenvalues);
+}
+
+template <typename index_type_t, typename value_type_t>
+void cupy_aux(raft::resources const& handle,
```

Comment:
Please don't prefix code in raft with `cupy_`. That's just bad form. This code is meant for RAFT, and we should treat it as such.